### PR TITLE
[Rust] Write dictionary data

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/io/writer.rs [hunk 1]
```diff
@@ -15,27 +15,80 @@
...
 use crate::io::object_writer::ObjectWriter;
 use crate::{Error, Result};
 
-/// FileWriter writes Arrow Table to a file.
+pub async fn write_manifest(writer: &mut ObjectWriter, manifest: &mut Manifest) -> Result<usize> {
+    // Write dictionary values.
+    let max_field_id = manifest.schema.max_field_id().unwrap_or(-1);
+    for field_id in 1..max_field_id + 1 {
+        if let Some(field) = manifest.schema.mut_field_by_id(field_id) {
+            if field.data_type().is_dictionary() {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/io/writer.rs:44]: This runs as a BFS over the nested schema.  Added a todo item to provide an `Iterator` over scheam.fields.

File: rust/src/io/writer.rs [hunk 2]
```diff
@@ -15,27 +15,80 @@
...
+                if let Some(dict_info) = field.dictionary.as_mut() {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/io/writer.rs:45]: lemme looking into it.

File: rust/src/io/writer.rs [hunk 3]
```diff
@@ -15,27 +15,80 @@
...
+                    if let Some(value_arr) = dict_info.values.as_ref() {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/io/writer.rs:46]: yea, was trying to make `format` module to be data structure only.

File: rust/src/io/writer.rs [hunk 4]
```diff
@@ -15,27 +15,80 @@
...
+                            _ => {
+                                return Err(Error::IO(format!(
+                                    "Does not support {} as dictionary value type",
+                                    value_arr.data_type()
+                                )));
+                            }
+                        };
+                        dict_info.offset = pos;
+                        dict_info.length = value_arr.len();
+                    } else {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/io/writer.rs]: fixed.

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**